File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/interview/P241_DifferentWaysToAddParentheses.java
package com.dsa.leetcode.labuladong.interview;

import java.util.*;

/**
 * @since 2024/8/3 22:32
 */
public class P241_DifferentWaysToAddParentheses {

    Map<String, List<Integer>> map = new HashMap<>();

    public List<Integer> diffWaysToCompute(String expression) {
        if (map.containsKey(expression)) {
            return map.get(expression);
        }

        List<Integer> res = new LinkedList<>();
        for (int i = 0, len = expression.length(); i < len; i++) {
            char c = expression.charAt(i);
            if (c == '+' || c == '-' || c == '*') {
                List<Integer> leftList = diffWaysToCompute(expression.substring(0, i));
                List<Integer> rightList = diffWaysToCompute(expression.substring(i + 1));
                for (Integer left : leftList) {
                    for (Integer right : rightList) {
                        if (c == '+') {
                            res.add(left + right);
                        } else if (c == '-') {
                            res.add(left - right);
                        } else {
                            res.add(left * right);
                        }
                    }
                }
            }
        }

        if (res.isEmpty()) {
            res.add(Integer.parseInt(expression));
        }
        map.put(expression, res);
        return res;
    }

    public static void main(String[] args) {
        String exp = "2*3-4*5";
        P241_DifferentWaysToAddParentheses p241 = new P241_DifferentWaysToAddParentheses();
        System.out.println(p241.diffWaysToCompute(exp));

    }

}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/interview/P241_DifferentWaysToAddParentheses.py
from typing import List


class Solution:
    def __init__(self):
        self.map = {}

    def diffWaysToCompute(self, expression: str) -> List[int]:
        if expression in self.map:
            return self.map[expression]

        res = []
        if expression.isdigit():
            res.append(int(expression))
            return res
        for i in range(len(expression)):
            c = expression[i]
            if c in ('+', '-', '*'):
                left_list = self.diffWaysToCompute(expression[:i])
                right_list = self.diffWaysToCompute(expression[i + 1:])
                for left in left_list:
                    for right in right_list:
                        if c == '+':
                            res.append(left + right)
                        elif c == '-':
                            res.append(left - right)
                        else:
                            res.append(left * right)
        self.map[expression] = res
        return res


if __name__ == '__main__':
    exp = "2*3-4*5"
    solu = Solution()
    print(solu.diffWaysToCompute(exp))
